targets/TARGET_RDA/TARGET_UNO_91H/trng_api.c

```diff
@@ -0,0 +1,98 @@
+/*
+ *  Hardware entropy collector for the UNO_91H, using RDA's TRNG
+ *
+ *  Copyright (C) 2006-2018, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#if defined(DEVICE_TRNG)
+
+#include "cmsis.h"
+#include "trng_api.h"
+#include "us_ticker_api.h"
+
+#define rRNG_CLKGATE        (RDA_SCU->CLKGATE0)
+#define rRNG_CLKGEN         (RDA_DMACFG->dma_func_ctrl)
+#define rRNG_INTSTATE       (RDA_DMACFG->dma_int_out)
+#define rTRNG_CTRL          (RDA_RNG->TCTRL)
+#define rTRNG_DATA0         (RDA_RNG->TD0)
+#define rTRNG_DATA1         (RDA_RNG->TD1)
+#define rPRNG_DATA          (RDA_RNG->PD)
+
+/** trng_get_byte
+ *  @brief Get one byte of entropy from the RNG, assuming it is up and running.
+ *  @param obj TRNG obj
+ *  @param pointer to the hardware generated random byte.
+ */
+static void trng_get_byte(trng_t *obj, unsigned char *byte )
+{
+    if (0U == obj->byte_idx) {
+        uint32_t start_time = us_ticker_read();
+        /* Delay for PRNG data ready, generate 4-byte per 32us */
+        while ((us_ticker_read() - start_time) <= 32);
+        obj->trng_data0 = rPRNG_DATA;
```

Comment:
A couple of more questions:
- Why are we making a copy of the register containing the random data?
- Why are we using the PRNG register? My impression was that it contains PRNG data if the TRNG is not on.
- Why do we assume that the TRNG is on, why can't we check it?

Reply:
rPRNG_DATA will update over time, so copy it.
We will enable TRNG in trng_init, TRNG can'r generate enough bits, so we use it as seed to produce PRNG data.
I have added init flag and check it in trng_get_bytes.

Reply:
Yes it will update over time, but if it is random it doesn't really matter when you read it if sufficient amount of time has past since the last read. But if you make a copy, then you have a value lying around in memory that were used as an input to a crypto library and depending on the design of the system this might increase the attack surface.

Reply:
i will get data from register directly.